mlir/lib/Dialect/Linalg/Transforms/ElementwiseOpFusion.cpp

```diff
@@ -1100,6 +1100,174 @@ class FoldPadWithProducerReshapeOpByExpansion
+        padOp.getConstantPaddingValue(), padOp.getNofold());
+    return success();
+  }
+
+private:
+  ControlFusionFn controlFoldingReshapes;
+};
+
+/// Pattern to fold a tensor.collapse_shape op with its producer tensor.pad op
+/// by bubbling the collapse_shape before the pad.
+struct FoldReshapeWithProducerPadOpByCollapsing
+    : public OpRewritePattern<tensor::CollapseShapeOp> {
+
+  FoldReshapeWithProducerPadOpByCollapsing(MLIRContext *context,
+                                           ControlFusionFn foldReshapes,
+                                           PatternBenefit benefit = 1)
+      : OpRewritePattern<tensor::CollapseShapeOp>(context, benefit),
+        controlFoldingReshapes(std::move(foldReshapes)) {}
+
+  LogicalResult matchAndRewrite(tensor::CollapseShapeOp collapseOp,
+                                PatternRewriter &rewriter) const override {
+    tensor::PadOp padOp = collapseOp.getSrc().getDefiningOp<tensor::PadOp>();
+
+    if (!padOp)
+      return failure();
+
+    if (!padOp->hasOneUse())
+      return failure();
+
+    if (!controlFoldingReshapes(&collapseOp.getSrcMutable())) {
+      return rewriter.notifyMatchFailure(collapseOp,
+                                         "fusion blocked by control function");
+    }
+
+    Value constantPaddingValue = padOp.getConstantPaddingValue();
+    if (!constantPaddingValue) {
+      return rewriter.notifyMatchFailure(
+          collapseOp, "cannot fold with non-constant padding value");
+    }
+
+    SmallVector<ReassociationIndices> reassociations =
+        collapseOp.getReassociationIndices();
+    SmallVector<OpFoldResult> low = padOp.getMixedLowPad();
+    SmallVector<OpFoldResult> high = padOp.getMixedHighPad();
+
+    for (auto [idx, reInd] : llvm::enumerate(reassociations)) {
+      if (reInd.size() > 1) {
+        for (auto dimIdx : reInd) {
+          if (!isConstantIntValue(low[dimIdx], 0) ||
+              !isConstantIntValue(high[dimIdx], 0)) {
+            return failure();
+          }
+        }
+      }
+    }
+
+    SmallVector<OpFoldResult> newLow, newHigh;
+    for (auto [idx, reInd] : llvm::enumerate(reassociations)) {
+      newLow.push_back(low[reInd[0]]);
+      newHigh.push_back(high[reInd[0]]);
+    }
+
+    Location loc = collapseOp.getLoc();
+    auto resultType = collapseOp.getResultType();
+
+    ArrayRef<int64_t> finalShape = collapseOp.getResultType().getShape();
+    SmallVector<int64_t> collapsedShape(finalShape.begin(), finalShape.end());
+    for (auto [inDimIdx, reInd] : llvm::enumerate(reassociations)) {
+      OpFoldResult l = low[reInd[0]];
+      OpFoldResult h = high[reInd[0]];
+      if (!isConstantIntValue(l, 0) || !isConstantIntValue(h, 0)) {
+        auto mixedSize =
+            tensor::getMixedSize(rewriter, loc, padOp.getSource(), reInd[0]);
+        auto dimSize = getConstantIntValue(mixedSize);
+        assert(dimSize.has_value() && "Expected static dimension");
+        collapsedShape[inDimIdx] = *dimSize;
```

Comment:
There should be no need to use `tensor::getMixedSize`, since we only need the static size. `tensor::getMixedSize` will insert new `tensor.dim` ops in the IR, which we don't want. Instead you can get the static size directly from the shape of the pad op source:
```suggestion
        collapsedShape[inDimIdx] = padOp.getSourceType().getShape()[reInd[0]];
```